Initial code review for main

salrashid123 opened the pull request:
Code review for GCP IAM auditing script.

Code will use Workspace domain delegation api to impersonate each user and for each user print which project are unassociated with the cloud org.

Cidan commented on main.go:
```diff
@@ -0,0 +1,393 @@
+		})
+		if err != nil {
+			glog.Errorf("Error creating CRM credentials for  user %s,  %v", u.PrimaryEmail, err)
+			return []*cloudresourcemanager.Organization{}, err
```
This should be `return nil, err`.

Cidan commented on main.go:
```diff
@@ -0,0 +1,393 @@
+		crmService, err = cloudresourcemanager.NewService(ctx, option.WithTokenSource(ts))
+		if err != nil {
+			glog.Errorf("Error creating CRM Service for  user %s,  %v", u.PrimaryEmail, err)
+			return []*cloudresourcemanager.Organization{}, err
```
This should be `return nil, err`.

Cidan commented on main.go:
```diff
@@ -0,0 +1,393 @@
+		})
+		if err != nil {
+			glog.Errorf("Error creating CRM credentials for  user %s,  %v", u.PrimaryEmail, err)
+			return []*cloudresourcemanager.Organization{}, err
```
This should be `return nil, err`.

Cidan commented on main.go:
```diff
@@ -0,0 +1,393 @@
+		crmService, err = cloudresourcemanager.NewService(ctx, option.WithCredentials(cred))
+		if err != nil {
+			glog.Errorf("Error creating CRM Service for  user %s,  %v", u.PrimaryEmail, err)
+			return []*cloudresourcemanager.Organization{}, err
```
This should be `return nil, err`.

Cidan commented on main.go:
```diff
@@ -0,0 +1,393 @@
+			glog.Fatalf("Error in rate limiter for user %s %v", u.PrimaryEmail, err)
+			return err
+		}
+		if ctx.Err() != nil {
```
Why do you need to check for error on the context? Shouldn't line 132 take care of this?

salrashid123 replied:
removed,
(you know, problems with finding code on reddit...)

Cidan commented on main.go:
```diff
@@ -0,0 +1,393 @@
+		return nil
+	}); err != nil {
+		glog.Errorf("Error iterating visible organizations for user %s %v", u.PrimaryEmail, err)
+		return []*cloudresourcemanager.Organization{}, err
```
This should be `return nil, err`.

Cidan commented on main.go:
```diff
@@ -0,0 +1,393 @@
+		crmService, err = cloudresourcemanager.NewService(ctx, option.WithTokenSource(ts))
+		if err != nil {
+			glog.Errorf("Error creating CRM Service for  user %s,  %v", u.PrimaryEmail, err)
+			return []*cloudresourcemanager.Project{}, err
```
This should be `return nil, err`.

Cidan commented on main.go:
```diff
@@ -0,0 +1,393 @@
+		crmService, err = cloudresourcemanager.NewService(ctx, option.WithCredentials(cred))
+		if err != nil {
+			glog.Errorf("Error creating CRM Service for  user %s,  %v", u.PrimaryEmail, err)
+			return []*cloudresourcemanager.Project{}, err
```
This should be `return nil, err`.

Cidan commented on main.go:
```diff
@@ -0,0 +1,393 @@
+	}
+
+	projects := make([]*cloudresourcemanager.Project, 0)
+	// do not apply a list filter here; we need to audit all projects
```
Clarify this comment, it should be a proper statement with capitalization and punctuation.

salrashid123 replied:
done, i've left that filter in so the reader/user can refine the projects list if they want to (i.,e they can see what bit to modify later if they choose)

Cidan commented on main.go:
```diff
@@ -0,0 +1,393 @@
+			glog.Errorf("Error in rate limiter for user %s %v", u.PrimaryEmail, err)
+			return err
+		}
+		if ctx.Err() != nil {
```
You shouldn't need to check the context error here, the error returned from limiter should work.

Cidan commented on main.go:
```diff
@@ -0,0 +1,393 @@
+	noFilter := ""
+
+	req := crmService.Projects.List().Filter(noFilter).PageSize(maxPageSize)
+	if err := req.Pages(ctx, func(page *cloudresourcemanager.ListProjectsResponse) error {
```
Don't split your `if err := func(); err != nil {` statements across multiple lines. This makes it hard to read/understand the flow, instead do:

```go
err := req.Pages(ctx, ...
...
)

if err != nil {
...
}
```

Cidan commented on main.go:
```diff
@@ -0,0 +1,393 @@
+	defer glog.Flush()
+	ctx := context.Background()
+
+	// configure API rate limits
```
Fix this comment the same as above.

Cidan commented on main.go:
```diff
@@ -0,0 +1,393 @@
+	}
+
+	if *serviceAccountFile != "" {
+		var err error
```
Remove this, use assignment from function:

```go
svcAccountJSONBytes, err := ...
```

salrashid123 replied:
i can't do the direct assignment since i need to see if the svcaccount file was specified or not first.  (i.,e if you use impersonation, the svcaccount file argument isn't set)

Cidan replied:
I'm talking about the `var err error` line here -- it's not clear to me why you are allocating it before hand. You can remove this, and make line 242 `:=` instead.

salrashid123 replied:
oh, ok. will fix in net push

salrashid123 replied:
hm. maybe i can't,

i've already defined `svcAccountJSONBytes`  as the global var, i can't use a := here  in that line

```golang
	if *serviceAccountFile != "" {
		var err error
		svcAccountJSONBytes, err = ioutil.ReadFile(*serviceAccountFile)
		if err != nil {
			glog.Fatal(err)
		}
	}
```

i could

```
	if *serviceAccountFile != "" {
		bt, err := ioutil.ReadFile(*serviceAccountFile)
		if err != nil {
			glog.Fatal(err)
		}
      svcAccountJSONBytes = bt
	}
```
but i don't thnk that buys much

Cidan replied:
Nah, it doesn't. Okay, just leave it. Not ideal, but since this is more "script" like rather than a real program, slightly less worried about globals.

Cidan commented on main.go:
```diff
@@ -0,0 +1,393 @@
+		}
+	}
+
+	// now initialize the Cloud Asset API
```
Fix these comments as in the above.

Cidan commented on main.go:
```diff
@@ -0,0 +1,393 @@
+		if err != nil {
+			glog.Fatal(err)
+		}
+	} else {
```
Remove this else and move the code block to the level above the else block -- Fatal in the first block halts execution.

Cidan commented on main.go:
```diff
@@ -0,0 +1,393 @@
+		}
+	}
+
+	// use asset-api API to recall/find the Organization for provided organizationID.  We don't really need to do this section
```
Fix this comment as in the above. While not required, try to keep comments under chars long (long lines for code are ok though).

Cidan commented on main.go:
```diff
@@ -0,0 +1,393 @@
+		allOrganizations[fmt.Sprintf("organizations/%s", orgName)] = response
+	}
+
+	// use asset-ap API to recall/find the projects under the provided organizationID.
```
Fix these comments as in the above.

Cidan commented on main.go:
```diff
@@ -0,0 +1,393 @@
+
+	glog.V(20).Infoln("      Getting Users")
+
+	// initialize the Workspace admin client first
```
Fix these comments as in the above.

Cidan commented on main.go:
```diff
@@ -0,0 +1,393 @@
+			glog.Fatal(err)
+		}
+	}
+	// if you want to narrow the search to a subset of users, apply a filter
```
Fix this comment as in the above.

Cidan commented on main.go:
```diff
@@ -0,0 +1,393 @@
+
+	req := crmService.Organizations.Search(&cloudresourcemanager.SearchOrganizationsRequest{Filter: noFilter, PageSize: maxPageSize})
+
+	if err := req.Pages(ctx, func(page *cloudresourcemanager.SearchOrganizationsResponse) error {
```
Do not split your if statements into new lines, see the other comment below about this.

Cidan commented on go.mod:
```diff
@@ -0,0 +1,13 @@
+module main
+
+go 1.15
```
Can you upgrade this to go 1.17? 1.15 is over a year old.

salrashid123 replied:
done...

..interesting, 1.17 changes splits up go.mod requires sections now

Cidan commented on main.go:
```diff
@@ -0,0 +1,393 @@
+
+	organizations := make([]*cloudresourcemanager.Organization, 0)
+
+	// do not apply a  filter here; we need to audit all projects
```
Clarify this comment, it should be a proper statement with capitalization and punctuation.

Cidan commented on main.go:
```diff
@@ -0,0 +1,393 @@
+		}
+		for _, u := range r.Users {
+			glog.V(20).Infof("      Found User: %s", u.PrimaryEmail)
+			// just use goroutines to print out the visibility
```
Clarify this comment, it should be a proper statement with capitalization and punctuation.

Cidan commented:
Just one small comment left on the `var err error` line

tbpg commented on README.md:
```diff
@@ -1 +1,221 @@
+
+`Bob` creates a project `D` in Cloud Org `OtherDomain`. `Bob` grants IAM permissions to `Alice `to have access to project `D`.  If `Alice` accepts the invite from `Bob` to use project `D`, `Alice` now has access to a project outside of the primary cloud org domain.
+
+In another situation, Alice may have created a project early on before the domain `MyDomain` was created.  Later on, an org admin will have to identify and [migrate projects](https://cloud.google.com/resource-manager/docs/project-migration) into the new domain, `MyDomain`.  
```
```suggestion
In another situation, `Alice` may have created a project early on before the domain `MyDomain` was created.  Later on, an org admin will have to identify and [migrate projects](https://cloud.google.com/resource-manager/docs/project-migration) into the new domain, `MyDomain`.  
```

tbpg commented on README.md:
```diff
@@ -1 +1,221 @@
+
+This script will help identify which projects a user in a domain may have been granted access to external projects or organizations.
+
+Either way, please review [Restricting project visibility for users](https://cloud.google.com/resource-manager/docs/access-control-org#restricting_visibility) in combination with [VPC-SC](https://cloud.google.com/vpc-service-controls/docs/overview)
```
```suggestion
Either way, please review [Restricting project visibility for users](https://cloud.google.com/resource-manager/docs/access-control-org#restricting_visibility) in combination with [VPC-SC](https://cloud.google.com/vpc-service-controls/docs/overview).
```

tbpg commented on README.md:
```diff
@@ -1 +1,221 @@
+
+1. Enumerate all projects within a given cloud org
+2. Iterate all users within a Workspace domain tied to the cloud org
+3. For each user, use _user impersonation_ to identify which organizations he/she has access to.
```
```suggestion
3. For each user, use _user impersonation_ to identify which organizations they have access to.
```

tbpg commented on README.md:
```diff
@@ -1 +1,221 @@
+1. Enumerate all projects within a given cloud org
+2. Iterate all users within a Workspace domain tied to the cloud org
+3. For each user, use _user impersonation_ to identify which organizations he/she has access to.
+4. For each user, use _user impersonation_ to identify which projects he/she has access to.
```
```suggestion
4. For each user, use _user impersonation_ to identify which projects they have access to.
```

tbpg commented on README.md:
```diff
@@ -1 +1,221 @@
+2. Iterate all users within a Workspace domain tied to the cloud org
+3. For each user, use _user impersonation_ to identify which organizations he/she has access to.
+4. For each user, use _user impersonation_ to identify which projects he/she has access to.
+5. If the user has access to any other organization thats not in the subject org in #1, print that external org
```
```suggestion
5. If the user has access to any other organization that's not in the subject org in #1, print that external org
```

tbpg commented on README.md:
```diff
@@ -1 +1,221 @@
+3. For each user, use _user impersonation_ to identify which organizations he/she has access to.
+4. For each user, use _user impersonation_ to identify which projects he/she has access to.
+5. If the user has access to any other organization thats not in the subject org in #1, print that external org
+6. If the user has access to any other project thats not in the set in #1, print that external project
```
```suggestion
6. If the user has access to any other project that's not in the set in #1, print that external project
```

tbpg commented on README.md:
```diff
@@ -1 +1,221 @@
+
+This script will:
+
+1. Enumerate all projects within a given cloud org
```
Nit: periods at the end for these.

tbpg commented on README.md:
```diff
@@ -1 +1,221 @@
+export DOMAIN_ADMIN=`gcloud config get-value core/account`
+
+# acquire the organization ID
+$ gcloud organizations list
```
Seems funny to put the `$` when this is already `bash`.

salrashid123 replied:
oh, its not necessary, i was just trying to show thats' run from a shell

i'll remove it

tbpg commented on README.md:
```diff
@@ -1 +1,221 @@
+
+Navigate to the Google Workspace Console and enable [Domain Wide Delegation](https://developers.google.com/admin-sdk/directory/v1/guides/delegation)
+
+Enter the `$SA_CLIENT_ID` for the service account and these _precise_  scopes
```
Sad there isn't a `gcloud` command for these steps.

salrashid123 replied:
naw, the dwd configuration is really a workspace feature (not in scope of GCP yet)

tbpg commented on main.go:
```diff
@@ -0,0 +1,399 @@
+
+/*
+
+Enumerate all projects and organizations Workspace users may have access to outside
```
Move this to be a package command and start this comment as "Command iam_visibility_audit ..."?

https://github.com/golang/go/wiki/CodeReviewComments#package-comments

tbpg commented on main.go:
```diff
@@ -0,0 +1,399 @@
+	return projects, nil
+}
+
+func main() {
```
This is totally minor, but I like to put `func main` at the top of `main.go` so it's clear where the entrypoint is. Doesn't really matter.

tbpg commented on cmd/main.go:
```diff
@@ -0,0 +1,398 @@
+	assetpb "google.golang.org/genproto/googleapis/cloud/asset/v1"
+)
+
+var (
```
Do these all need to be globals? Not a big deal, but especially for `wg` I prefer to pass explicit arguments to make it easier to see the flow of the program. See the other comments for some ideas of what functions might be helpful to do this.

salrashid123 replied:
moved to local variables.  i left the `limiter` as global...i suppose i can move that as parameter too...WDYT?

tbpg commented on cmd/main.go:
```diff
@@ -0,0 +1,398 @@
+
+	// Initialize the Cloud Asset API.
+	// This api is used to find the projects in an ORG
+	var assetClient *asset.Client
```
Consider moving all of the client creation into separate functions. Then, you can initialize the `client` variables directly using those functions. You also have the benefit of being able to return early instead of needing an `else`. Finally, it makes `func main` shorter and easier to understand because the name of the function says what it does.

Note the separate functions should return errors and not `glog.Fatal` directly.

tbpg commented on cmd/main.go:
```diff
@@ -0,0 +1,398 @@
+	// Use the Cloud Asset API to recall/find the Organization for provided organizationID.
+	// We don not really need to do this section since the the orgID was provided in the input argument.
+	// We are doing this as a type of input validation with the asset-api to search/narrow the organization.
+	assetType := "cloudresourcemanager.googleapis.com/Organization"
```
This request and collection of responses in the `for` loop seems like it should (1) be a separate function and (2) return the list rather than set a global variable.

salrashid123 replied:
generalized the organiation and projectlist functions by asset type

tbpg commented on cmd/main.go:
```diff
@@ -0,0 +1,398 @@
+	}
+
+	// Use the Cloud Asset API to recall/find the Organization for provided organizationID.
+	// We don not really need to do this section since the the orgID was provided in the input argument.
```
```suggestion
	// We do not really need to do this section since the the orgID was provided in the input argument.
```

tbpg commented on cmd/main.go:
```diff
@@ -0,0 +1,398 @@
+		allOrganizations[fmt.Sprintf("organizations/%s", orgName)] = response
+	}
+
+	// Use the Cloud Asset API to recall/find the projects for provided organizationID.
```
Same thing here -- this could be a separate function `findProjects` or something.

salrashid123 replied:
as above, collapsed to one function

tbpg commented on cmd/main.go:
```diff
@@ -0,0 +1,398 @@
+
+	// Initialize the Workspace Admin client
+	// This client will be used to find users in a given Cloud Identity/Workspace domain
+	var adminService *admin.Service
```
Initialize in a separate function like `assetClient`.

tbpg commented on cmd/main.go:
```diff
@@ -0,0 +1,398 @@
+	//  https://developers.google.com/admin-sdk/directory/v1/guides/search-users
+	//	eg. searchFilter := "isAdmin=false"
+
+	searchFilter := ""
```
`searchFilter` could be an argument to a function that does the actual query. :smile:

salrashid123 replied:
done, added in as parameters

tbpg commented on cmd/main.go:
```diff
@@ -0,0 +1,398 @@
+			wg.Add(1)
+			go getOrganizations(ctx, *u)
+			wg.Add(1)
+			go getProjects(ctx, *u)
```
Might want to separate querying the user and finding the relevant users. Right now, I don't know where the user info gets printed. I'd expect the printing to happen in `func main`.

salrashid123 replied:
done, i added in go routines and channels....(its a first for me using channels here so PTAL careful look!)

salrashid123 replied:
removed channels so closing this out infavor of the refactor

tbpg commented on cmd/main.go:
```diff
@@ -0,0 +1,398 @@
+			go getOrganizations(ctx, *u)
+			wg.Add(1)
+			go getProjects(ctx, *u)
+			time.Sleep(time.Duration(*delay) * time.Millisecond)
```
What is the purpose of the delay?

salrashid123 replied:
The project.list is throttled at something like 4pqs (eg, cloudresourcemanager.projects.list is globally at ~4)

i tried using 
`"golang.org/x/time/rate"`  set to 3 or lower but even then needed a 1s delay per user.

i ran this for about 1800 users which took 20mins or so.  This combo seems to work fine and plays nice with quota avilability vs consumption at steady-state

tbpg reviewed (commented):
This seems like a fantastic example to have on hand that

1. Shows how to do something useful and necessary in and of itself.
2. Is an example of using a few different services.

Thank you!

I left some refactoring-type comments. Overall, it might be easier to follow if some chunks of code were in separate functions. Then, the `main` function could orchestrate, with all of the technical details hidden away.

salrashid123 commented:
ok @tbpg , refactored a fair amount there and added in a channel to emit the findings back.

i also added in an option to shard quota between projects

tbpg commented on go.mod:
```diff
@@ -0,0 +1,27 @@
+
+go 1.17
+
+require (
```
Looks like you need a `go mod tidy`. I'm surprised by the two `require` blocks.

salrashid123 replied:
i think its go 1.17?

```$ go version
go version go1.17.1 linux/amd64

$ go mod tidy

$ more go.mod 
module github.com/salrashid123/iam_visibility_audit

go 1.17

require (
	cloud.google.com/go/asset v1.0.1
	github.com/golang/glog v1.0.0
	golang.org/x/net v0.0.0-20211007125505-59d4e928ea9d
	golang.org/x/oauth2 v0.0.0-20211005180243-6b3c2da341f1
	golang.org/x/time v0.0.0-20210723032227-1f47c861a9ac
	google.golang.org/api v0.58.0
	google.golang.org/genproto v0.0.0-20211007155348-82e027067bd4
)

require (
	cloud.google.com/go v0.94.1 // indirect
	github.com/golang/groupcache v0.0.0-20200121045136-8c9f03a8e57e // indirect
	github.com/golang/protobuf v1.5.2 // indirect
	github.com/google/go-cmp v0.5.6 // indirect
	github.com/googleapis/gax-go/v2 v2.1.1 // indirect
	go.opencensus.io v0.23.0 // indirect
	golang.org/x/sys v0.0.0-20210917161153-d61c044b1678 // indirect
	golang.org/x/text v0.3.6 // indirect
	google.golang.org/appengine v1.6.7 // indirect
	google.golang.org/grpc v1.40.0 // indirect
	google.golang.org/protobuf v1.27.1 // indirect
)
```

tbpg commented on go.mod:
```diff
@@ -0,0 +1,27 @@
+module main
```
```suggestion
module github.com/salrashid123/iam_visibility_audit
```

tbpg commented on main.go:
```diff
@@ -0,0 +1,475 @@
+	assetpb "google.golang.org/genproto/googleapis/cloud/asset/v1"
+)
+
+var ()
```
Delete.

tbpg commented on main.go:
```diff
@@ -0,0 +1,475 @@
+	go func() {
+		for {
+			msg, ok := <-ch
+			if ok {
```
```suggestion
		for msg := range ch {
```

When you `range` over a channel, it stops when the channel gets closed. See https://tour.golang.org/concurrency/4.

tbpg commented on main.go:
```diff
@@ -0,0 +1,475 @@
+	// getOrganizations(), and getProjects() function calls in the succeeding stanza.
+	ch := make(chan *userAccess)
+	done := make(chan bool)
+	go func() {
```
Move this after the `range allUsers` loop and there is no need for this goroutine. That also means you can get rid of `done`.

tbpg commented on main.go:
```diff
@@ -0,0 +1,475 @@
+		}
+		return asset.NewClient(ctx, option.WithTokenSource(ts))
+
+	} else {
```
No need for the `else {` because the `if` ends in a `return`. https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow is relevant. Applies in a couple other spots, too.

tbpg commented on main.go:
```diff
@@ -0,0 +1,475 @@
+		// see https://pkg.go.dev/google.golang.org/api@v0.58.0/cloudresourcemanager/v1#SearchOrganizationsRequest
+		noFilter := ""
+		wg.Add(1)
+		go getOrganizations(ctx, ch, &wg, limiter, noFilter, impersonateAccount, svcKeyCred, *u)
```
Consider doing something like this:

```go
go func() {
    wg.Add(1)
    defer wg.Done()
    userOrgs, err := getOrganizations(ctx, ...)
    if err != nil { ... }
    userProjects := getProjects(ctx, ...)
    if err != nil { ... }
    ch <- userAccess{...} // No need for the Error field anymore.
}()
```

The organizations and projects calls don't happen concurrently, but that's OK because there is enough else going on concurrently it shouldn't matter in the end. Benefits:

* You don't have to pass the channel to `getOrganization` or `getProjects` -- they don't need to know they're being called concurrently at all; they just return the values they were asked for.
* You also don't have to pass `wg`.
* The organization and project info for each user is all printed next to each other, not whatever order the runtime scheduler decides.

salrashid123 replied:
ok, so i'm a bit confused on this and the previous bit

so, the first go routine launches and watis on the `ch` to feed it stuff.   When that channel is closed it fulfills `done`.

The `getOrganizations` `getProjects` both run in the background and finds the projects per and feeds into `ch`.  Once every user is iterated,  the waitgroup if fulfilled/empty.  

The next step waits on `done` being complete (because i need to wait until the stuff pending in the channel is empty, right?


the closest i can come up with is

```golang
ch := make(chan *userAccess)
done := make(chan bool)
go func() {
  for msg := range ch {
    for _, o := range msg.Organizations {
      glog.V(50).Infof("             User %s has Organization visibility to %s", msg.User.PrimaryEmail, o.Name)
      if _, ok := allOrganizations[o.Name]; !ok {
        glog.V(2).Infof("             User [%s] has external organization visibility to [%s](%s)", msg.User.PrimaryEmail, o.Name, o.DisplayName)
      }
    }
    for _, p := range msg.Projects {
      glog.V(50).Infof("             User %s has Project visibility to %s", msg.User.PrimaryEmail, p.ProjectId)
      if _, ok := allProjects[p.ProjectId]; !ok {
        glog.V(2).Infof("             User [%s] has external project visibility to [projects/%d](%s)", msg.User.PrimaryEmail, p.ProjectNumber, p.ProjectId)
      }
    }
  }
  done <- true
}()

	for _, u := range allUsers {
		impersonateAccount, svcKeyCred, err := getRandomServiceAccount(serviceAccounts, svcAccountKeys)
		if err != nil {
			glog.Errorf("Error finding valid serviceAccount %v", err)
			return
		}

		wg.Add(1)
		go func() {
			defer wg.Done()
			organizationFilter := ""
			projectFilter := ""
			userOrganizations, err := getOrganizations(ctx, limiter, organizationFilter, impersonateAccount, svcKeyCred, *u)
			if err != nil {
				glog.Errorf("Error finding Organizations for user %s %v", u.PrimaryEmail, err)
				return
			}
			userProjects, err := getProjects(ctx, limiter, projectFilter, impersonateAccount, svcKeyCred, *u)
			if err != nil {
				glog.Errorf("Error finding Projects for user %s %v", u.PrimaryEmail, err)
				return
			}
			ch <- &userAccess{
				User:          *u,
				Projects:      userProjects,
				Organizations: userOrganizations,
			}
		}()
		time.Sleep(time.Duration(*delay) * time.Millisecond)
	}


wg.Wait()
close(ch)
<-done
```

salrashid123 replied:
closing this since  i'm no longer using channels and refactored it anyway

tbpg commented on main.go:
```diff
@@ -0,0 +1,475 @@
+		//  https://pkg.go.dev/google.golang.org/api@v0.58.0/cloudresourcemanager/v1#ProjectsListCall.Filter
+		wg.Add(1)
+		go getProjects(ctx, ch, &wg, limiter, noFilter, impersonateAccount, svcKeyCred, *u)
+		time.Sleep(time.Duration(*delay) * time.Millisecond)
```
Could you use the `limiter` here rather than in the `getOrganizations` and `getProjects` functions? Or does it need finer granularity?

salrashid123 replied:
the `getOrganizations` and `getProjects` both uses the same api method so uses the same quota.

even with a limiter set to < 4qps, this always seemed to trip the api quota threhdolds..i did't quite undestand how thats happening (another quota that this trips?).  i'll test this out.  (i tried removing the sleep may times)

tbpg commented on main.go:
```diff
@@ -0,0 +1,475 @@
+	// We do not really need to do this section since the the orgID was provided in the input argument.
+	// We are also setting a default "emptyQuery" value here to return all projects.
+	// We are doing this as a type of input validation with the asset-api to search/narrow the organization.
+	emptyQuery := ""
```
Suggestion, feel free to ignore. `findResourcesByAssetType` only ever uses `emptyQuery`. Consider either documenting possible query values or removing that parameter entirely.

salrashid123 replied:
ok, added a comment on usage and an example of selecting on-pending deleted project ` queryFilter = "state=ACTIVE"`

tbpg commented on main.go:
```diff
@@ -0,0 +1,453 @@
+	// Launch goroutines for each user and find which projects and organizations they have access to
+	for _, u := range allUsers {
+
+		wg.Add(1)
```
Nit: Consider moving the declaration of `wg` to right before this loop.

tbpg commented on main.go:
```diff
@@ -0,0 +1,453 @@
+			// Find the projects and organizations that do NOT exist in the list of projects under the subject organization.
+			// The projects the user has access to that is not included in []allProjects should be outside the subject organization.
+			for _, o := range userOrganizations {
+				glog.V(50).Infof("             User %s has Organization visibility to %s", u.PrimaryEmail, o.Name)
```
```suggestion
				glog.V(50).Infof("             User [%s] has Organization visibility to %s", u.PrimaryEmail, o.Name)
```

tbpg commented on main.go:
```diff
@@ -0,0 +1,453 @@
+			}
+			glog.V(50).Infof("             User [%s] can see %d projects", u.PrimaryEmail, len(userProjects))
+			for _, p := range userProjects {
+				glog.V(50).Infof("             User %s has Project visibility to %s", u.PrimaryEmail, p.ProjectId)
```
```suggestion
				glog.V(50).Infof("             User [%s] has Project visibility to %s", u.PrimaryEmail, p.ProjectId)
```

tbpg commented on main.go:
```diff
@@ -0,0 +1,453 @@
+
+			// Find the projects and organizations that do NOT exist in the list of projects under the subject organization.
+			// The projects the user has access to that is not included in []allProjects should be outside the subject organization.
+			for _, o := range userOrganizations {
```
Any issue with interleaving log statements about different users? Might want a mutex. But, probably not a big deal.

salrashid123 replied:
nope, no issue with interleaving stuff,  its pretty easy for a user to sort in order from the output

tbpg commented on main.go:
```diff
@@ -0,0 +1,453 @@
+			glog.V(20).Infof("     Found projectID %s", projectID)
+			resourceList[projectID] = response
+		default:
+			return nil, fmt.Errorf(fmt.Sprintf("Error getting resources:  unknown assetType: %s", assetType))
```
No need for `fmt.Errorf(fmt.Sprintf(...))`, you can just do `fmt.Errorf(...)` directly.

```suggestion
			return nil, fmt.Errorf("error getting resources:  unknown assetType: %s", assetType)
```

Also, https://github.com/golang/go/wiki/CodeReviewComments#error-strings for the capital letter at the start.
